052-record-reconciled-versions.md

```diff
@@ -0,0 +1,186 @@
+- The Reconciler should update some value on reconcilliation startup without overriding the `strimzi.io/reconciled` annotation, the reason for this is, if for any reason the operator errors during reconcile it would incorrectly update `reconciled` to Y, despite not finishing to successful completion. So having a two stage commit mechanism here would give much more information to a user regarding whether it started reconcile versus whether it finished.
+
+## `Kafka` Custom Resource
+- The update to `strimzi.io/reconciling` would happen in the `initialStatus` [method](https://github.com/strimzi/strimzi-kafka-operator/blob/2a1fdf9d8695bb22a2bf977b0ba5414291530207/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java#L292)
+with `strimzi.io/reconciled` remaining unchanged, leaving it unset if unset, or leaving it as whatever it was set to previously if set.
+- The update to `strimzi.io/reconciled` would happen in the `createOrUpdate` [method](https://github.com/strimzi/strimzi-kafka-operator/blob/2a1fdf9d8695bb22a2bf977b0ba5414291530207/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java#L137) alongside the `Ready` update where the operator version would be updated into the field.
+- `KafkaReconciler` will also need to be modified to wait for all SPS to be updated, I believe in the `podsReady` [method](https://github.com/strimzi/strimzi-kafka-operator/blob/2a1fdf9d8695bb22a2bf977b0ba5414291530207/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaReconciler.java#L817C4-L817C4) where instead of checking the pods the operator will insteady wait for the SPS CR to have `strimzi.io/reconciled` set to the operator version signalling its completion.
+
+
+## `StrimziPodSet` Custom Resource
+- The update to `strimzi.io/reconciling` would happen in the `reconcile` [method](https://github.com/strimzi/strimzi-kafka-operator/blob/2a1fdf9d8695bb22a2bf977b0ba5414291530207/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/StrimziPodSetController.java#L382)
+with `strimzi.io/reconciled` remaining unchanged, leaving it unset if unset, or leaving it as whatever it was set to previously if set.
+- The update to `strimzi.io/reconciled` would happen in the `reconcile` [method](https://github.com/strimzi/strimzi-kafka-operator/blob/2a1fdf9d8695bb22a2bf977b0ba5414291530207/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/StrimziPodSetController.java#L407) alongside the finish of the status update where the operator version would be updated into the field.
+
+
+## Examples
+
+For the following example any key set to `null` counts as it not being set for explicitness and clarity
+
+### Fresh Install
+
+- On a fresh install, a user creates a Kafka CR:
+  ```
+  strimzi.io/reconciled: null
+  strimzi.io/reconciling: null
+  ```
```

Comment:
Would these actually exist? We don't expect the user to include them in their CR?

Reply:
added a clarifying comment